### DISCUSSION TITLE
Printing mdBook SRS to different files / Generating multi-file software artifacts

### DISCUSSION THREAD

**Discussion by BilalM04** [JacquesCarette/Drasil#3796] [Q&A]
Tickets to revisit after this discussion is resolved:
* #388 

<hr>

Currently, the way printing SRS documents are setup, it is only possible to print to a single file. Each example makes a call to the [`gen`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-gen/lib/Language/Drasil/Generate.hs#L37-L41) function in  [`Language.Drasil.Generate`](https://github.com/JacquesCarette/Drasil/blob/mdBookPrinter/code/drasil-gen/lib/Language/Drasil/Generate.hs), passing in the SRS, which is a `Document`:

https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-example/projectile/app/Main.hs#L19

Then, through a series of functions in [`Language.Drasil.Generate`](https://github.com/JacquesCarette/Drasil/blob/mdBookPrinter/code/drasil-gen/lib/Language/Drasil/Generate.hs), particularly [`prnt`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-gen/lib/Language/Drasil/Generate.hs#L42-L48), [`prntDoc`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-gen/lib/Language/Drasil/Generate.hs#L50-L61), [`printDoc'`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-gen/lib/Language/Drasil/Generate.hs#L63-L75), and [`writeDoc`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-gen/lib/Language/Drasil/Generate.hs#L94-L100),  a call to a format specific gen function is made:

https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-gen/lib/Language/Drasil/Generate.hs#L93-L101

These format specific gen functions ([`genHTML`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/HTML/Print.hs#L46-L49), [`genTeX`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/TeX/Print.hs#L41-L45), [`genJSON`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/JSON/Print.hs#L32-L37), [`genMD`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Markdown/Print.hs#L31-L33)) return a single `Doc`, which contains the entire SRS. This `Doc` is then printed to its respective SRS file.

The format specific gen functions receive a single `Document`, which contains all the SRS details. The `Document` it receives is in the format of the data type defined in [`Language.Drasil.Document`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-lang/lib/Language/Drasil/Document.hs#L61-L64), where the SRS contents is represented through `[Section]`. The format specific gen functions, using the [`makeDocument`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Printing/Import/Document.hs#L22-L27) function, then convert it to the `Document` data type defined in [`Language.Drasil.Printing.LayoutObj`](https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Printing/LayoutObj.hs#L9-L10), where the SRS content is represented through `[LayoutObj]`. The SRS printers then render these `LayoutObj`'s into a single `Doc`. `LayoutObj` is defined as follows:

https://github.com/JacquesCarette/Drasil/blob/bfcf92f09b07bd883ef2af438f300499d35305e0/code/drasil-printers/lib/Language/Drasil/Printing/LayoutObj.hs#L30-L43

---
The description above is how Drasil currently generates SRS documents (rather *my* understanding of it, please correct me if I'm wrong). The problem here is that the SRS is treated as a single document all throughout its journey from conception to being printed. 

After playing around with the code for printing, I found the general structure of the SRS contents (represented as `[LayoutObj]`) follows a nested structure. The top-level `LayoutObj`s in the list are `HDiv`s for each of the primary sections (closed dot in the TOC). The `HDiv`s contain `[LayoutObj]`, which is used to nest the sub-sections (open dot in the TOC). Similarly. the sections denoted by the square in the TOC are also nested. 

<details>

<summary>Here are the table of contents for reference</summary>

- Table of Contents
- Reference Material
  - Table of Units
  - Table of Symbol
  - Abbreviations and Acronyms
- Introduction
  - Purpose of Document
  - Scope of Requirements
  - Characteristics of Intended Reader
  - Organization of Document
- General System Description
  - System Context
  - User Characteristics
  - System Constraints
- Specific System Description
  - Problem Description
    - Terminology and Definitions
    - Physical System Description
    - Goal Statements
  - Solution Characteristics Specification
    - Assumptions
    - Theoretical Models
    - General Definitions
    - Data Definitions
    - Instance Models
    - Data Constraints
    - Properties of a Correct Solution
- Requirements
  - Functional Requirements
  - Non-Functional Requirements
- Traceability Matrices and Graphs
- Values of Auxiliary Constants
- References

</details>

From my understanding, some `HDiv`s, regardless of whether it is nested or not, should be printed into separate files. Whereas there are some `HDiv`s that do not need to be printed to a separate file, i.e. the ones containing block equations. However, this nested structure makes it difficult to separate the SRS into different sections. I can think of several "hacky" ways to go about it; however being new to Drasil and Haskell, I am unsure of a good design. 

---

**What is the *best* way to print the mdBook SRS into separate Documents?**

**Comment by JacquesCarette** [comment]
You're probably going to need to generalize this to (at least) return a 'set' of documents. It might be clever to put in a bit of abstraction now. I'm guessing that a Map from some reasonable set of 'key's to `Doc`s might be a good first implementation.

You might want to take a peek at `drasil-code` and `drasil-GOOL` (with some help from @B-rando1 ) as it deals with many files already, so its architecture may be something to learn from.

In any case: this is a sub-optimal design of the `gen` functions -- we've just never needed to generate multiple files for documents!

**Reply by balacij** [reply]
@BilalM04 has a [working prototype](https://github.com/JacquesCarette/Drasil/commit/0beb49d4ea735dd7868c2a2b305a440cfde31594) that breaks the SRS into various files, using headers (sec/subsec/subsubsec, in a `LayoutObj`)  to determine if some "block" of the SRS should be in its own file. However, we're running into issues with _references that reference things belonging to other files_.

### Discussion of References Issue

The approach takes a post-`SRSDecl` rendering approach to building a file structure. However, by the time a document's `LayoutObj`s are created, internal reference chunks are already created (and rendered within the `LayoutObj`), making the assumption that our internal references will all be within the same file.

So, the issue: links like `[TM:something](#TM:something)` from an `InstanceModels.md` section now need to be updated to include the files the links belong to (relative to the working file's directory). For example, if we have
```
A.md
b/B.md
b/c/C.md
```
and we want to link things from `B` or `C` in `A`, we can use absolute references. However, if we wanted to link to `A` or `B` from `C`, we would need to use relative references (e.g., `../../A.md#thing` and `../B.md#thing2`).

To proceed, we need to figure out which ID is defined in which file. Again, currently, that is done internally and automatically, before any sort of rendering.

Aside: Internally, IDs are really only used for areas where we _render a specific chunk_. So, we're really looking to find where we placed our 'rendered chunks' so that we can build proper 'chunk anchors' (references to the rendered chunks).

We need to:
1. figure out where each ID node (target/chunk rendering) is placed, and
2. adjust the references (anchors) to accommodate the directory structure and information found in (1).

To figure out where each ID node (target/chunk rendering) is placed, we can either:
* a. use the information that @BilalM04 currently excavates from `LayoutObj`s to also figure out which file will render which chunk (and hence be a target/ID node), or
* b. try to move reference building to the chunk rendering stage (caching the information as we go along).

Proceeding with (1) means that we will need to go over all `LayoutObj`s before finally rendering them into `Doc`s to retroactively 'fix' chunk anchors (i.e., sentences that contain a `Ref` node [below]).

https://github.com/JacquesCarette/Drasil/blob/864e1385e94f900d8035f605a0b170b9c439dc61/code/drasil-printers/lib/Language/Drasil/Printing/AST.hs#L55

However, this would mean a few things:
i. our pre-constructed references IDs will become a 'best guess' rather than 'truth,' and
ii. we would be performing string-search-and-replace-like operations on artifacts that we render.

Furthermore, in Markdown, IDs are normally (i.e., outside our use case of Markdown) automatically constructed from headers, usually taking the form `x-y-z-...`. In HTML, using IDs of that form are also [generally](https://medium.com/@somameeta/best-practices-for-naming-ids-in-html-elements-32a929b5f6bc) [good](https://stackoverflow.com/questions/880776/best-practice-for-naming-the-id-attribute-of-dom-elements) [practice](https://courses.cs.washington.edu/courses/cse154/17au/styleguide/html-css/naming-conventions-html.html). However, our currently generated IDs take on a form (i.e., `X:Y:Z`) more common to LaTeX.

I think I tend more towards solution (b) to figure out where each ID is defined because it will allow us to specialize IDs to renderers but also uses 'truth' (provided by the renderers) in a single pass. However, going down the route of solution (b) would mean that, when rendering `Doc`s, we need to know where each chunk anchor belongs to already. So, we would need to make sure that our links have a 'tree-like' structure and be very careful that we render files in proper order so that we can only write references to things that have been previously rendered.

However, @BilalM04 caught one issue with this approach: we often have sections that need to reference each other. For example, with assumptions, we carry both references to assumptions in various areas, and we carry a list of what areas reference each particular assumption alongside its definition.

A solution to this issue could involve us upgrading our `Doc`s. Right now, we use `pretty`'s `Doc` type to build documents. However, we might need to implement 'holes' (in our `Doc`s) that we can backfill later on. We can either do that by hacking the `Doc` data type ourselves (and perhaps the `pretty` package) or using text-based search-and-replace within `Doc`s.

Regarding the task of switching our renderers to ones that also define chunk anchors along the way, I built a prototype of what I think this _can_ look like for MdBook generation (note: it's rough) as part of building a prototype [stateful chunk renderer](https://github.com/balacij/drasil-artifacts/blob/main/src/Drasil/Artifacts.hs):

https://github.com/balacij/drasil-artifacts/blob/a0b0b12f0f7b591ed71de6d0756062928ad5da48/app/Main.hs#L47-L65

**Reply by balacij** [reply]
Note: in my prototype, I did _not_ yet build the `Doc` 'hole' feature out yet.

**Reply by balacij** [reply]
I think that document 'holes' are a reasonable feature to desire. It reminds me of LaTeX builds -- the reason why we run the LaTeX compiler multiple times is so that we can compile all of our floating lists and whatnot in early builds, so that later builds can make use of said information. However, that information is dependent on the fact that the latex _has not changed_.

**Reply by balacij** [reply]
In regard to #1025, would the stateful renderer (and `Doc`s with 'hole' support) discussed above be helpful? I think so... but I'm not confident.

Our current approach involves searching our rendered chunks to figure out what things references other things, which is fair, but it is disconnected from how we render chunks, which is ultimately what we care about. Additionally, this feature of figuring out what things reference other things in a rendered document is probably a useful feature for all kinds of documents, not just our SRS. So it might be something that we prefer at the document-renderer level, so that if others decide to use the document language to render chunks, the feature is already built. This also extends to us: if we choose to generate a VnV, this feature comes for cheap.

**Reply by balacij** [reply]
This all being said: I need to upgrade my prototype if I want to convince anyone, with at least the following:
1. Document holes.
2. A system for building chunk renderers that automagically build (and record) chunk anchors.

**Reply by smiths** [reply]
Cross-file references are an important problem to solve.  This won't be the last time they come up.  When we eventually generate design and testing documents, we will need plenty of traceability between the documents.  In particular, we will (in some sense) want everything to trace back to requirements.  For LaTeX documents I've used the `extern` package for this in the past, but we are using `hyperref` for our internal links.  A little googling suggests that [hyperref can do external links to other files](https://tex.stackexchange.com/questions/41539/does-hyperref-work-between-two-files), although we would need to do more research to find if this is the best approach.  Whatever solution we find for mdbooks will hopefully be consistent with cross-referencing between documents in other contexts in the future.

**Comment by smiths** [comment]
@BilalM04, I like your [description](https://github.com/JacquesCarette/Drasil/discussions/3796#discussion-6809343) above of how the SRS is generated.  I'm guessing that you deduced this from reading the code, or maybe it is documented somewhere.  If it isn't documented, we should talk about capturing the overview somewhere for the benefit of future Drasil developers.  I'll add this discussion to our Monday meeting (#3791).
